table/table_iterator.h

```diff
@@ -48,22 +49,65 @@ class TableIterator : public Iterator {
   bool Valid() const override { return iter_->Valid(); }
   void SeekToFirst() override { return iter_->SeekToFirst(); }
   void SeekToLast() override { return iter_->SeekToLast(); }
-  void Seek(const Slice& target) override { return iter_->Seek(target); }
+  void Seek(const Slice& target) override {
+      std::string seek_key_buf;
+      ROCKSDB_NAMESPACE::Status s = ROCKSDB_NAMESPACE::GetInternalKeyForSeek(target, options_->comparator,
+                                                                             &seek_key_buf);
+      assert(s.ok());
+      if (!s.ok()) {
+        status_ = s;
+      }
+      iter_->Seek(seek_key_buf);
+  }
   void SeekForPrev(const Slice& target) override {
-    return iter_->SeekForPrev(target);
+      std::string seek_key_buf;
+      ROCKSDB_NAMESPACE::Status s = ROCKSDB_NAMESPACE::GetInternalKeyForSeekForPrev(target, options_->comparator,
+                                                                                    &seek_key_buf);
+      if(!s.ok()) {
+          status_ = s;
+      }
+      iter_->SeekForPrev(target);
+  }
+
+  void Next() override {
+      iter_->Next();
+      init_key();
   }
-  void Next() override { return iter_->Next(); }
-  void Prev() override { return iter_->Prev(); }
-  Slice key() const override { return iter_->key(); }
+
+  void Prev() override {
+      iter_->Prev();
+      init_key();
+  }
+
+  Slice key() const override { return ikey->user_key;}
   Slice value() const override { return iter_->value(); }
-  Status status() const override { return iter_->status(); }
+  Status status() const override {
+      if (!status_.ok()) {
+          return status_;
+      }
+      return iter_->status();
+  }
   Status GetProperty(std::string /*prop_name*/,
                      std::string* /*prop*/) override {
     assert(false);
     return Status::NotSupported("TableIterator does not support GetProperty.");
   }
 
+  uint64_t SequenceNumber() {
```

Comment:
The public API is `Iterator` in rocksdb/includes/iterator.h, the `TableIterator` is an internal class that is not supposed to be publicly accessible. These `SequenceNumber`, and `type` APIs are not defined for `Iterator`.


We have added `ParseEntry` util methods to parse sequence number and type out of an internal key. You can have you own wrapper iterator class wrapping a RocksDB `Iterator`, and add the parsing logic in to the wrapper iterator.

Reply:
We need all of these infos separately and storing the parsed entry would be much more optimal. Otherwise for each and every call we would have to parse the key over and over again. All I am doing here is changing the definition of the table iter which would seek over the same user key and return user key when getKey() is called. I have added 2 more functions on top of iterator, I believe that should be fine, since JNI would be called only when TableIterator is initialized.

Reply:
@jowlyzhang TableIterator itself is a wrapper for InternalIterator why have 2 wrappers? Should we consider making this public?

Reply:
It's worth having 2 wrappers because it will be your own wrapper iterator that you don't need to check into RocksDB codebase, you own those logic and you can modify it however you want. Parsing an entry every time over and over again isn't computationally heavy or performance wise concerning, you can just wrap it in your own wrapper iterator. It's beset for the RocksDB `TableIterator` to just do the most vanilla things of iterating raw entries. Others wishing to add their own iterating constraints can just add on top of it.

Reply:
@jowlyzhang I have updated the code to add another jni functionality to convert internal Key to user key and vice versa